src/Groups/matrices/linear_centralizer.jl

```diff
@@ -0,0 +1,400 @@
+# Compute a generating set for every element in GL(n,q) and SL(n,q)
+# generators for GL(n,q) and SL(n,q) are described in:  Taylor, D. E., Pairs of Generators for Matrix Groups. I, The Cayley Bulletin, 3 (1987)
+# generators for centralizer of unipotent elements are described in: Giovanni De Franceschi, Centralizers and conjugacy classes in finite classical groups, arXiv:2008.12651
+
+
+########################################################################
+#
+# Orders
+#
+########################################################################
+
+function _GL_order(n::Int, q::fmpz)
+   res = q^div(n*(n-1),2)
+   for i in 1:n res *= (q^i-1) end
+   return res
+end
+
+_GL_order(n::Int, F::Ring) = _GL_order(n, order(F))
+
+function _SL_order(n::Int, q::fmpz)
+   res = q^div(n*(n-1),2)
+   for i in 2:n res *= (q^i-1) end
+   return res
+end
+
+_SL_order(n::Int, F::Ring) = _SL_order(n, order(F))
+
+########################################################################
+#
+# Centralizer in GL
+#
+########################################################################
+
+
+# returns as matrices
+function _gens_for_GL(n::Int, F::Ring)
+   n !=1 || return [matrix(F,1,1,[primitive_element(F)])]
+   if order(F)==2
+      h1 = identity_matrix(F,n)
+      h1[1,2] = 1
+      h2 = zero_matrix(F,n,n)
+      h2[1,n] = 1
+      for i in 1:n-1 h2[i+1,i] = -1 end
+      return h1,h2
+   else
+      h1 = identity_matrix(F,n)
+      h1[1,1] = primitive_element(F)
+      h2 = zero_matrix(F,n,n)
+      h2[1,1] = -1
+      h2[1,n] = 1
+      for i in 1:n-1 h2[i+1,i] = -1 end
+      return h1,h2
+   end
+end
+
+# returns as matrices
+# does the same as above with F = F[x]/(f), but every entry is replaced by a diagonal join of D corresponding blocks
+# ASSUMPTION: deg(f) > 1
+function _gens_for_GL_matrix(f::PolyElem, n::Int, F::Ring; D=1)
+   C = companion_matrix(f)
+   CP = evaluate(_centralizer(f),C)            # matrix of maximal order in the centralizer of the companion matrix
+
+   if n==1 return [diagonal_join([CP for i in 1:D])] end
+   h1 = identity_matrix(F,n*degree(f)*D)
+   insert_block!(h1,diagonal_join([CP for i in 1:D]),1,1)
+   h2 = zero_matrix(F,n*degree(f)*D,n*degree(f)*D)
+   for i in 1:(n-1)*degree(f)*D h2[i+degree(f)*D,i]=-1 end
+   for i in 1:degree(f)*D
+      h2[i,i]=-1
+      h2[i,i+(n-1)*degree(f)*D]=1
+   end
+   return h1,h2
+end
+
+
+# V = vector of integers of the dimensions of Jordan blocks
+# return the generators for the centralizers of the diagonal join of unipotent Jordan blocks of dimensions V
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+function _centr_unipotent(F::Ring, V::AbstractVector{Int}; isSL=false)
+   n = sum(V)
+   _lambda = gen(F)  # yes, gen(F) is correct; we don't need a primitive element in this case
```

Comment:
According to the documentation:
>  gen(a::FqFiniteField)

>  Return the generator of the finite field. Note that this is only guaranteed to be a multiplicative generator if
  the finite field is generated by a Conway polynomial automatically.

In this case I need just `gen(F)`, i.e. a generator of `F` as a field. When I need a generator of `F*` as cyclic group, I use `primitive_element(F)`.

Reply:
Since we had this discussion also for GAP:
The notation in the literature is not uniform.
In some books (and also in GAP), the term "primitive element" is used for any element that generates the given field (over the given subfield), and any generator of the multiplicative group of a field is called "primitive root".

Reply:
I can rename `primitive_element` to `primitive_root`.

Reply:
> I can rename `primitive_element` to `primitive_root`.

I think this will be helpful.
As far as I see, `gen` means "primitive element" in the sense of a generator of a field extension:
The docstring for `gen(L::SimpleNumField)` mentions explicitly that the result is "the canonical primitive element of L over K".